### PR TITLE
[BugFix] Fix query partition compensate predicates for mv rewrite

### PR DESCRIPTION
**Pull request by LiShuMing**
Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java [hunk 1]
```diff
@@ -279,10 +279,17 @@ private boolean isRollupAggregate(List<ScalarOperator> mvGroupingKeys, List<Scal
...
+            if (operator instanceof BinaryPredicateOperator) {
+                BinaryPredicateOperator binaryPredicateOperator = ((BinaryPredicateOperator) operator);
+                if (!binaryPredicateOperator.getBinaryType().isEqual()) {
+                    continue;
+                }
+
+                if (operator.getChild(0).isColumnRef() && operator.getChild(1).isConstantRef()) {
+                    equalsPredicateColumns.add(operator.getChild(0));
+                } else if (operator.getChild(1).isColumnRef() && operator.getChild(0).isConstantRef()) {
+                    equalsPredicateColumns.add(operator.getChild(1));
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java:291]: is it not canonized?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 2]
```diff
@@ -1199,58 +1251,63 @@ private ScalarOperator addJoinDerivePredicate(
...
+                    logMVRewrite(mvRewriteContext, "derive join's extra predicate failed, mv join type: {}, " +
+                            "query join type: {}", mvJoinOp, queryJoinOp);
                     return null;
                 }
                 ScalarOperator derivedPredicate = derivedPredicateOpt.get();
                 if (!derivedPredicate.equals(ConstantOperator.TRUE)) {
                     derivedPredicates.add(derivedPredicate);
                 }
-                List<ColumnRefOperator> leftJoinColumns = joinDeriveContext.getLeftJoinColumns();
+                // add lfet table's not null constrains
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:1303]:
```suggestion
                // add left table's not null constraints
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java [hunk 3]
```diff
@@ -14,76 +14,89 @@
...
+
+    // Like equal/range/residual predicates above, but all equal/range/residual predicates
+    // contain all valid or invalid predicates which invalid predicates are not used for
+    // rewrite but are used for other checks.
+    private final PredicateSplitter inValidSplitter;
+
+    private PredicateSplit(PredicateSplitter validSplitter,
+                           PredicateSplitter inValidSplitter) {
+        this.validSplitter = validSplitter;
+        this.inValidSplitter = inValidSplitter;
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/PredicateSplit.java:36]: it is hard to understand here

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 86 / 95 (90.53%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|57|64|89.06%|[465, 490, 823, 830, 872, 873, 874]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|19|21|90.48%|[100, 101]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewriteContext.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 91 / 103 (88.35%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/Log4jConfig.java|12|18|66.67%|[198, 199, 200, 201, 207, 211]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|27|31|87.10%|[830, 872, 873, 874]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|25|27|92.59%|[100, 101]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/RewriteContext.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|19|19|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewriteContext.java|4|4|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request closed without merging**